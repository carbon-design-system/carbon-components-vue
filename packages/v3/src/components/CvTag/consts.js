export const tagKinds = [
  'red',
  'magenta',
  'purple',
  'blue',
  'cyan',
  'teal',
  'green',
  'gray',
  'cool-gray',
  'warm-gray',
  'high-contrast',
];
const CvTagConsts = { tagKinds };
export default CvTagConsts;
