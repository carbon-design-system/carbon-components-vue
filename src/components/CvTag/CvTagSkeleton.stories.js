import { storyParametersObject } from '../../global/storybook-utils';
import CvTagSkeleton from './CvTagSkeleton.vue';

export default {
  title: 'Component/CvTagSkeleton',
  component: CvTagSkeleton,
};

const template = `<cv-tag-skeleton v-bind="args"></cv-tag-skeleton>`;
const Template = (args, { argTypes }) => {
  return {
    props: Object.keys(argTypes),
    components: { CvTagSkeleton },
    template,
    setup() {
      return { args };
    },
  };
};

export const Default = Template.bind({});
Default.parameters = storyParametersObject(
  Default.parameters,
  template,
  Default.args
);

export const Small = Template.bind({});
Small.args = {
  small: true,
};
Small.parameters = storyParametersObject(
  Small.parameters,
  template,
  Small.args
);
