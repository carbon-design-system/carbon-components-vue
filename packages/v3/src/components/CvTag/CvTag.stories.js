import CvTag from './CvTag';
import { tagKinds } from './consts.js';
import { action } from '@storybook/addon-actions';
import {
  sbCompPrefix,
  storyParametersObject,
} from '../../global/storybook-utils';

export default {
  title: `${sbCompPrefix}/CvTag`,
  component: CvTag,
  argTypes: {
    kind: { control: { type: 'select', options: tagKinds } },
  },
};

const template = `<CvTag @remove="onRemove" v-bind="args" />`;
const Template = (args, { argTypes }) => {
  return {
    props: Object.keys(argTypes),
    components: { CvTag },
    template,
    setup() {
      return {
        args,
        onRemove: action('remove'),
      };
    },
  };
};

export const Default = Template.bind({});
Default.args = {
  kind: 'red',
  label: 'This is a tag',
  filter: false,
};
Default.parameters = storyParametersObject(
  Default.parameters,
  template,
  Default.args
);

export const Filter = Template.bind({});
Filter.args = {
  kind: 'teal',
  label: 'This is a tag',
  filter: true,
};
Filter.parameters = storyParametersObject(
  Filter.parameters,
  template,
  Filter.args
);

export const Skeleton = Template.bind({});
Skeleton.args = {
  skeleton: true,
};
Skeleton.parameters = storyParametersObject(
  Skeleton.parameters,
  template,
  Skeleton.args
);
