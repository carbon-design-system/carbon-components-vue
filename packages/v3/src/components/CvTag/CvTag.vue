<template>
  <div v-if="skeleton" :class="tagClasses" />
  <div v-else :class="tagClasses" role="listitem" :title="title">
    <span :class="`${carbonPrefix}--tag__label`">
      {{ label }}
    </span>
    <button
      v-if="filter"
      :class="`${carbonPrefix}--tag__close-icon`"
      :aria-label="clearAriaLabel"
      @click.stop.prevent="onRemove"
      :disabled="disabled"
    >
      <Close16 />
    </button>
  </div>
</template>

<script>
import { computed } from 'vue';
import { carbonPrefix } from '../../global/settings';
import { tagKinds } from './consts';
import Close16 from '@carbon/icons-vue/es/close/16';

export default {
  name: 'CvTag',
  components: { Close16 },
  props: {
    clearAriaLabel: { type: String, default: 'Clear filter' },
    /**
     * disabled by property or if skeleton
     */
    disabled: Boolean,
    /**
     * label to be used in the CvTag component
     */
    label: { type: String, required: true },
    /**
     * kind of the CvTag
     */
    kind: {
      type: String,
      default: 'red',
      validator(val) {
        return tagKinds.includes(val);
      },
    },
    /**
     * If filter is true, the CvTag will include a remove button on the right side, which on a click, will emit the 'remove' event.
     */
    filter: {
      type: Boolean,
      default: false,
    },
    /**
     * skeleton used when loading
     */
    skeleton: Boolean,
  },
  emits: [
    /**
     * Triggers when the clear button is pressed on a filter CvTag
     */
    'remove',
  ],
  setup(props, { emit }) {
    const title = computed(() => {
      return props.filter ? props.clearAriaLabel : null;
    });

    const tagClasses = computed(() => {
      const classes = [
        `${carbonPrefix}--tag`,
        `${carbonPrefix}--tag--${props.kind}`,
      ];

      if (props.skeleton) {
        classes.push(`${carbonPrefix}--skeleton`);
      } else {
        if (props.filter) {
          classes.push(`${carbonPrefix}--tag--filter`);
        }

        if (props.disabled) {
          classes.push(`${carbonPrefix}--tag--disabled`);
        }
      }
      return classes;
    });

    const onRemove = () => {
      if (!props.disabled) {
        emit('remove');
      }
    };

    return {
      carbonPrefix,
      title,
      tagClasses,
      onRemove,
    };
  },
};
</script>
