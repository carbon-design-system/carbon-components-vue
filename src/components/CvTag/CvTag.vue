<template>
  <div v-if="skeleton" :class="tagClasses"></div>
  <div v-else :class="tagClasses" role="listitem" :title="title">
    <cv-svg
      v-if="!skeleton && renderIcon"
      :svg="renderIcon"
      :class="`${carbonPrefix}--tag__custom-icon`"
    />
    <span :class="`${carbonPrefix}--tag__label`">
      {{ label }}
    </span>
    <button
      v-if="filter"
      :class="`${carbonPrefix}--tag__close-icon`"
      :aria-label="clearAriaLabel"
      :disabled="disabled || null"
      @click.stop.prevent="onRemove"
    >
      <Close16 />
    </button>
  </div>
</template>

<script>
import { computed } from 'vue';
import { carbonPrefix } from '../../global/settings';
import { tagKinds } from './consts';
import Close16 from '@carbon/icons-vue/es/close/16';
import CvSvg from '../CvSvg/_CvSvg.vue';

export default {
  name: 'CvTag',
  components: { CvSvg, Close16 },
  props: {
    /** aria label to use for the x icon for the filter tag */
    clearAriaLabel: { type: String, default: 'Clear filter' },
    /** disabled by property or if skeleton */
    disabled: Boolean,
    /** label to be used in the CvTag component */
    label: { type: String, required: true },
    /** kind of the CvTag */
    kind: {
      type: String,
      default: undefined,
      validator(val) {
        return tagKinds.includes(val);
      },
    },
    /**
     * If filter is true, the CvTag will include a remove button on the right side, which on a click, will emit the
     * 'remove' event.
     */
    filter: {
      type: Boolean,
      default: false,
    },
    /** skeleton used when loading */
    skeleton: Boolean,
    /** tag size small */
    small: Boolean,
    /** Optional prop to render a custom icon. \@carbon/icons-vue icon, href, svg or symbol */
    renderIcon: {
      type: [String, Object],
      default: undefined,
    },
  },
  emits: [
    /**
     * Triggers when the clear button is pressed on a filter CvTag
     */
    'remove',
  ],
  setup(props, { emit }) {
    const title = computed(() => {
      return props.filter ? props.clearAriaLabel : null;
    });

    const tagClasses = computed(() => {
      const classes = [`cv-tag ${carbonPrefix}--tag`];

      if (props.small) classes.push(`${carbonPrefix}--tag--sm`);

      if (props.skeleton) {
        classes.push(`${carbonPrefix}--skeleton`);
      } else {
        if (props.kind) {
          classes.push(`${carbonPrefix}--tag--${props.kind}`);
        }

        if (props.filter) {
          classes.push(`${carbonPrefix}--tag--filter`);
        }

        if (props.disabled) {
          classes.push(`${carbonPrefix}--tag--disabled`);
        }
      }
      return classes;
    });

    const onRemove = () => {
      if (!props.disabled) {
        emit('remove');
      }
    };

    return {
      carbonPrefix,
      title,
      tagClasses,
      onRemove,
    };
  },
};
</script>
