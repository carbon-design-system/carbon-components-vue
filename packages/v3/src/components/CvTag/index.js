import CvTag from './CvTag';
import CvTagConsts from './consts';

export { CvTag, CvTagConsts };
export default CvTag;
