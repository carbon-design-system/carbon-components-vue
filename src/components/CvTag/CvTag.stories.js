import CvTag from './CvTag.vue';
import { tagKinds } from './consts.js';
import { action } from '@storybook/addon-actions';
import {
  sbCompPrefix,
  storyParametersObject,
} from '../../global/storybook-utils';
import {
  DataDefinition20,
  Bee20,
  Carbon20,
  Watson20,
  IbmCloud20,
  EdtLoop20,
  IbmSecurity20,
} from '@carbon/icons-vue';

const icons = {
  DataDefinition20,
  Bee20,
  Carbon20,
  Watson20,
  IbmCloud20,
  EdtLoop20,
  IbmSecurity20,
  undefined,
};

export default {
  title: `${sbCompPrefix}/CvTag`,
  component: CvTag,
  argTypes: {
    kind: { control: 'select', options: tagKinds },
    renderIcon: { control: { type: 'select', options: Object.keys(icons) } },
  },
};

const template = `<cv-tag @remove="onRemove" v-bind="args" />`;
const Template = (argsIn, { argTypes }) => {
  let args = { ...argsIn, renderIcon: icons[argsIn.renderIcon] };
  return {
    props: Object.keys(argTypes),
    components: { CvTag },
    template,
    setup() {
      return {
        args,
        onRemove: action('remove'),
      };
    },
  };
};

export const Default = Template.bind({});
Default.args = {
  label: 'Hugo Award winner',
  filter: false,
};
Default.parameters = {
  controls: {
    exclude: ['remove'],
  },
};
Default.parameters = storyParametersObject(
  Default.parameters,
  template,
  Default.args
);

export const Filter = Template.bind({});
Filter.args = {
  label: 'Filter House',
  filter: true,
  kind: 'high-contrast',
};
Filter.parameters = {
  controls: {
    exclude: ['remove'],
  },
};
Filter.parameters = storyParametersObject(
  Filter.parameters,
  template,
  Filter.args
);

export const Skeleton = Template.bind({});
Skeleton.args = {
  skeleton: true,
  label: '',
};
Skeleton.parameters = {
  controls: {
    include: ['small'],
  },
};
Skeleton.parameters = storyParametersObject(
  Skeleton.parameters,
  template,
  Skeleton.args
);

export const TagIcon = Template.bind({});
TagIcon.args = {
  renderIcon: 'DataDefinition20',
  label: 'Steampunk',
};
TagIcon.parameters = {
  controls: {
    include: ['renderIcon'],
  },
};
TagIcon.parameters = storyParametersObject(
  TagIcon.parameters,
  template,
  TagIcon.args
);
