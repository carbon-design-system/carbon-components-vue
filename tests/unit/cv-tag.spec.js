import { shallowMount } from "@vue/test-utils";
import CvTag from "@/components/cv-tag/cv-tag.vue";

describe("CvTag.vue", () => {
  it("renders props.label when passed", () => {
    const label = "new label";
    const wrapper = shallowMount(CvTag, {
      propsData: { label },
    });

    expect(wrapper.props("label")).toBe(label);
    expect(wrapper.text()).toMatch(label);
  });

  it("checks props.label is required", () => {
    expect(CvTag.props.label.required).toBe(true);
  });

  it("checks props.kind has default", () => {
    expect(CvTag.props.kind.default).toBeDefined();
  });

  it("matches snapshot", () => {
    const label = "I am a tag";
    const kind = "beta";
    const wrapper = shallowMount(CvTag, {
      propsData: { label, kind },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
