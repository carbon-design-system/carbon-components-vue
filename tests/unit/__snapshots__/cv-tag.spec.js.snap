// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvTag.vue matches snapshot 1`] = `<span class="cv-tag bx--tag bx--tag--beta">I am a tag</span>`;
